sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc

```diff
@@ -0,0 +1,373 @@
+that they would like to use. The user might want to express their function
+using SIMD types because they need to use intrinsics or inline ASM, have legacy
+code that uses SIMD types, or are more familiar with that programming model.
+This extension adds support for an execution model in which SIMD parallelism is
+expressed using SIMD types, with clear delineation between SPMD and SIMD code regions through an invocation interface also introduced by the extension.
+
+=== Defining a SIMD Function
+
+Uniform arguments are represented by scalars, and varying arguments by explicit
+vector types.  The explicit vector types are based on the SIMD classes
+from [Parallelism TS2](http://wg21.link/N4808), and support all of the functions
+operating on SIMD types as defined there. Additionally, each of the functions
+defined in Sections 4.17.5, 4.17.6 and 4.17.7 of the SYCL 2020 specification
+are overloaded to accept SIMD classes; these overloads operate element-wise on
+the contents of the SIMD class.
+
+[source, c++]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+namespace simd_abi {
+template <class T, int N>
+using native_fixed_size = /* implementation-defined */;
+} // namespace simd_abi
+
+template <class T, int N>
+using simd = std::experimental::simd<T, simd_abi::native_fixed_size<N>>;
+
+template <class T, int N>
+using simd_mask = std::experimental::simd_mask<T, simd_abi::native_fixed_size<N>>;
+
+} // namespace experimental
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+----
+
+We introduce a new `native_fixed_size` ABI which combines elements of the
+existing `native` and `fixed_size` ABIs defined in Parallelism TS2;
+implementations are expected to use the most efficient representation
+of a `simd<T, N>` type (e.g. using multiple registers rather than an array).
+We expect implementations of this ABI to build on Clang's `ext_vector_type`,
+rather than specifying the register types directly.
+
+The value of `N` for each `simd` argument to a SIMD function must be the same,
+and represents the sub-group size of the calling kernel.  If a function accepts
+`simd` arguments with multiple `N` values, it cannot be invoked from a SPMD
+kernel and the implementation must issue a diagnostic in this case.
+
+The example below shows a simple SIMD function that scales all elements of
+an 8-wide SIMD type `x` by a scalar value `n`.
+
+[source, c++]
+----
+sycl::ext::oneapi::experimental::simd<float, 8> scale(sycl::ext::oneapi::experimental::simd<float, 8> x, float n)
+{
+  return x * n;
+}
+----
+
+=== Invoking a SIMD Function
+
+The `invoke_simd` free function invokes a SIMD function using all work-items
+in a `sub_group`.  The `invoke_simd` interface marshals data between the
+SPMD context of the calling kernel and the SIMD context of the callee,
+converting arguments and return values between scalar and SIMD types as
+appropriate.
+
```

Comment:
Per offline discussion with @rolandschulz, the spec might need to require that the caller of `invoke_simd` should have explicitly set subgroup size - i.e. bear the `[[sycl::reqd_sub_group_size(N)]]` attribute. 

For given SPMD caller `foo` and SIMD callee `bar` the SG size could be deduced from vector arguments of the `bar` (unless `bar` signature contains uniform types only, in which case SG size does not matter), but that would present problems for SPMD callers of the `foo`, which must know if it has fixed SG size, esp. if the caller does not see `foo` body.